Issue#266 (manual save thumbnail)

rostyslavvnahornyi opened the pull request.

rostyslavvnahornyi committed:
feat: allowed setting scene thumbnail from capture share modal

rostyslavvnahornyi committed:
add: firestore rules for storage

rostyslavvnahornyi commented:
@kfarr By the way, I did this task without the firebase plugin, all at once on the frontend code

kfarr commented on src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx:
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
     screenshotEl.setAttribute('screentock', 'takeScreenshot', true);
+    try {
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+      const aframeScene = document.getElementById('screentock-destination');
```
`aframeScene` is actually an img HTML element, can we rename to a more accurate ID for readibility such as `screentockImgElement`

(note to self: screentock-destination is created by toolbar component as img element that receives canvas drawing)

kfarr commented on src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx:
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
+    try {
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+      const aframeScene = document.getElementById('screentock-destination');
+      const canvas = await html2canvas(aframeScene);
```
- this new `canvas` element is created from an existing image using html2canvas
- I think this is not needed and even if it is, not sure we need a package to do this?
- Instead, see below

kfarr commented on src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx:
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
+      resizedCanvas.height = 240;
+      const context = resizedCanvas.getContext('2d');
+
+      context.drawImage(canvas, 0, 0, 320, 240);
```
Instead of converting img element to canvas via html2canvas and then drawing in another canvas; in theory we can pass the screentock-destination img directly. (`drawImage()` should accept `HTMLImageElement`)

So this could be:
`     context.drawImage(aframeScene, 0, 0, 320, 240);`

kfarr commented on src/components/scenegraph/Toolbar.js:
```diff
@@ -147,11 +147,15 @@ export default class Toolbar extends Component {
 
       // we want to save, so if we *still* have no sceneID at this point, then create a new one
-      if (!currentSceneId || !!doSaveAs) {
+      if (!currentSceneId || (!!doSaveAs && !isCurrentUserTheSceneAuthor)) {
```
I don't understand why these changes are related to thumbnail upload (#266)?

kfarr commented on src/components/scenegraph/Toolbar.js:
```diff
@@ -147,11 +147,15 @@ export default class Toolbar extends Component {
         currentSceneId = await generateSceneId(this.props.currentUser.uid);
+        if (!doSaveAs) {
+          window.location.hash = `#/scenes/${currentSceneId}.json`;
+        }
```
I don't understand why these changes are related to thumbnail upload (#266)?

kfarr commented on src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx:
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
 import { useAuthContext } from '../../../contexts';
 import Toolbar from '../../scenegraph/Toolbar';
+import { db, storage } from '../../../services/firebase';
+import html2canvas from 'html2canvas';
```
can we try to not have this dependency? see below

kfarr committed:
remove html2canvas dependency

kfarr committed:
maintain aspect ratio and cover all destination canvas

kfarr commented:
To integrate:
- [x] remove altered Save As... logic from Toolbar.js -- is this related to another ticket / PR?
- [x] allow firing of this event with a button per figma design (may require separating the code such that it can be fired on demand and not with download as...)
- [x] remove firing of this event when just downloading a file
- [x] security rules check: as a user, I should not be able to upload a new scene thumbnail to a scene to which I am not an owner
- [x] current scene for URL path is determined by wrong method (url bar instead of metadata)
- [x] move the new STREET.utils from index.html into json-utils (and create ticket https://github.com/3DStreet/3dstreet/issues/390 to convert other global functions to STREET.utils)
- [x] add success / error notify messages
- [x] investigate save as error
- [ ] merge!


picture1:
![image](https://github.com/3DStreet/3dstreet-editor/assets/470477/4f19b967-7e8b-465a-b893-ac287a73e722)

kfarr committed:
revert save as logic changes

see https://github.com/3DStreet/3dstreet-editor/issues/307 instead

kfarr committed:
fix linebreaks

kfarr committed:
remove html2canvas junk

kfarr committed:
dist bump

kfarr committed:
imperfect restyling and adding button

button still needs to be reassigned

kfarr committed:
move thumbnail upload to its own function

kfarr committed:
don't fire when downloading images

kfarr committed:
fix storage security rules

- fix only allow write if user is author of scene
- fix limit access to large files unless logged in

kfarr committed:
fix thumbnail saving id issue

- bug with old logic to use url to determine thumbnail path
- unify getting scene id and title into STREET.utils global

kfarr committed:
update json utils from 3dstreet core

kfarr committed:
move new utils to json-utils.js file

also try new global object

kfarr committed:
add success / error messages

kfarr committed:
update scene metadata with sceneId when saving

fixes https://github.com/3DStreet/3dstreet-editor/issues/307
